### PR TITLE
update SmoothQuant algorithm with folding choice

### PR DESCRIPTION
**Pull request by xin3he**
## Type of Change

update Smoothquant algorithm with folding choice

## Description
```
#  By default, folding args: {IPEX: False (True if version<2.1), ONNX RT: False, Stock PT: True} 
folding=False: Allow inserting mul to update the input distribution and not absorbing.
folding=True: Only allow inserting mul with the input scale that can be absorbed into the last layer. 
```
IPEX can fuse inserted mul automatically and `folding=False` is recommanded.
For stock PT, `folding=True` mode won't introduce performance issue for int8 model while `folding=False` mode will insert mul and quant/dequant pair for int8 model, just for accuracy verification.

## Expected Behavior & Potential Risk

UT pass

## How has this PR been tested?

locally test with latest ipex and stock pt.

## Dependency Change?

Need latest ipex2.1 for ipex backend.

### CODE REVIEW

**Review by hshen14 — APPROVED** [review]
LGTM

File: neural_compressor/algorithm/smooth_quant.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def __init__(self, alpha=0.5):
         # scales_per_op: True, each op will have an individual scale, mainly for accuracy
         #                False, ops with the same input will share a scale, mainly for performance
         self.alpha = alpha
+        self.folding = False
```
Comment (PenghuiCheng) [inline neural_compressor/algorithm/smooth_quant.py:49]: Can we add a descriptor for this argument?

Comment (PenghuiCheng) [reply neural_compressor/algorithm/smooth_quant.py:49]: I mean we should add a descriptor in docs.

Comment (xin3he) [reply neural_compressor/algorithm/smooth_quant.py:49]: I will add document in next PR.

**Review by PenghuiCheng — APPROVED** [review]

**Review by yintong-lu — APPROVED** [review]

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 2]
```diff
@@ -16,28 +16,25 @@
...
 from ...utils import logger
 
 
 def model_forward(model, dataloader, iters):
+    from ..pytorch import pytorch_forward_wrapper
     try:
         cnt = 0
         for idx, (input, label) in enumerate(dataloader):
-            output = model(input)
+            output = pytorch_forward_wrapper(model, input)
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:30]: is any issue here? I'm not too fond of coupling this with INC's other module here.  Add  another "device" parameter should be enough

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 3]
```diff
@@ -101,6 +98,111 @@ def quant_dequant_x(x, num_bits=8):
...
+        # update weight w/ QDQ
+        weith_qdq = quant_dequant_w(self.module)
+        self.module.weight = torch.nn.Parameter(weith_qdq)
+
+
+class SQLinearWrapper(torch.nn.Module):
+    def __init__(self, module, input_scale, input_minmax, dtype=torch.quint8):
+        super().__init__()
+        self.input_scale = input_scale
+        self.dtype = dtype
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:163]: I prefer to follow the 'sym'/‘asym’ terminology which is more popular

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 4]
```diff
@@ -101,6 +98,111 @@ def quant_dequant_x(x, num_bits=8):
...
+        return X
+
+    def _calculate_qparams(self, input_scale, input_minmax, dtype=torch.quint8):
+        # calculate scale and zero_point
+        if dtype == torch.quint8:
+            quant_min, quant_max = 0, 255
+        min_val = torch.min(input_minmax[0] * input_scale)
+        max_val = torch.max(input_minmax[1] * input_scale)
+        # work when min_val bigger than zero.
+        min_val_neg = torch.min(min_val, torch.zeros_like(min_val))
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:184]: this function shared some common code with the function quant_dequant_x.  Could we unify them into one?

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 5]
```diff
@@ -101,6 +98,111 @@ def quant_dequant_x(x, num_bits=8):
...
+        scale = (max_val_pos - min_val_neg) / float(quant_max - quant_min)
+        scale = torch.max(scale, torch.tensor([torch.finfo(torch.float32).eps]))
+        zero_point = quant_min - torch.round(min_val_neg / scale).to(torch.int)
+        zero_point = torch.clamp(zero_point, quant_min, quant_max)
+        return scale, zero_point
+
+    def _get_weight_scale(self):
+        # get weight scale and zero_point
+        from torch.ao.quantization.observer import default_per_channel_weight_observer
+        obs = default_per_channel_weight_observer()
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:195]: the code is quite simple, we could copy the code

File: neural_compressor/adaptor/torch_utils/smooth_quant.py (hunk 2 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/smooth_quant.py:30]: It's more stable than the older one with more condition judgement. Reusing table function is recommended, I think

File: neural_compressor/adaptor/torch_utils/smooth_quant.py (hunk 4 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/smooth_quant.py:184]: I don't think so. It's not a quant_dequant function, I'm calculating a scale from observerd minmax list, not input.

File: neural_compressor/adaptor/torch_utils/smooth_quant.py (hunk 3 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/smooth_quant.py:163]: Here I want to introduce int4 with dtype later. It's not used now, and input is always asym for better accuracy.

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 6]
```diff
@@ -255,11 +350,11 @@ def _calibrate(self, absorb_to_layer, calib_iter, save_input_output=False):
             return {}
         hook_modules_input_output = {}
         for name in self.hook_layer_names:
-            hook_modules_input_output[name] = self._get_module(name)
+            hook_modules_input_output[name] = get_module(self.model, name)
         self._add_observer(hook_modules, hook_modules_input_output)
         self._dump_min_max(calib_iter=calib_iter)
         self._remove_observer()
-        return self.input_maxes
+        return self.input_abs_maxes
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:357]: nice catch, thank you

File: neural_compressor/adaptor/pytorch.py [hunk 7]
```diff
@@ -1251,18 +1253,55 @@ def smooth_quant(self, model, dataloader, calib_iter, tune_cfg=None, alpha=0.5,
...
+
+        if self.performance_only:
+            q_model = model
+        else:
+            try:
+                q_model = copy.deepcopy(model)
+            except Exception as e:  # pragma: no cover
+                logger.warning("Fail to deep copy the model due to {}, inplace is used now.".format(
+                    repr(e)))
+                q_model = model
```
[the remark this answers is not in the document]
Comment (xin3he) [reply neural_compressor/adaptor/pytorch.py:1270]: It's INC design, we shouldn't overwrite the fp32_model after quantization unless performance_only.

[the remark this answers is not in the document]
Comment (xin3he) [reply neural_compressor/adaptor/pytorch.py:1271]: You are right, we should replace the fp32 model after pre-optimization, will remove deepcopy here.

File: neural_compressor/adaptor/torch_utils/smooth_quant.py [hunk 8]
```diff
@@ -128,13 +259,19 @@ def __init__(self, model, dataloader, traced_model=None):
         self.input_values = {}
         self.output_values = {}
         self.input_maxes = {}
+        self.input_mines = {}
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/smooth_quant.py:262]: mins?

Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/smooth_quant.py:262]: I see, it's a typo error.

**Comment by chensuyue** [comment]
Please fix CI issue.

**Review by changwangss — APPROVED** [review]
